### ISSUE TITLE
Concern about putting `FetchContent_MakeAvailable(…)` in a `block()`

### ISSUE THREAD

**Issue by dabrahams** [bemanproject/exemplar#61]
https://github.com/beman-project/exemplar/blob/727a5c62a60b426c859e9905c136b9012a82fa83/CMakeLists.txt#L29C1-L30C1

I don't do it for reasons explained [here](https://github.com/hylo-lang/CMakeModules/blob/6577fca51495e82a94e76a8b70113c9e6b98f539/FetchHyloDependency.cmake#L12), FWIW.

**Comment by camio** [comment]
What @dabrahams linked to:

```
# Not using block() here because FetchContent_MakeAvailable typically causes dependency-specific
# global variables to be set, and I'm not sure to what extent they may be needed
```

There is no such thing as "global" variables in CMake. There are CACHE entries, which `block()` will not impact, and normal variables which have a scope that can only be propagated above one level (via `set(... PARENT_SCOPE)`). I don't understand the issue here.

**Comment by dabrahams** [comment]
Fine, it should have said

```
# Not using block() here because FetchContent_MakeAvailable typically causes dependency-specific
# variables to be set in the parent scope, and I'm not sure to what extent they may be needed
```

**Comment by camio** [comment]
If there are dependency-specific variables required to use a dependency, they should be CACHE variables. I can't fathom why normal variables would be used for this purpose.

**Comment by dabrahams** [comment]
I can't fathom why you're talking about CACHE variables. I'm talking about variables that will be set by `FetchContent_MakeAvailable`; just look at the docs for `find_package` and note all the variables it says will be set.  Only 4 of them are documented there to be cache variables.

**Comment by dabrahams** [comment]
But you know, feel free to close this issue if my concerns are misplaced.

**Comment by wusatosi** [comment]
Closing this issue as it appears to be stale

**Comment by camio** [comment]
I'm reopening this issue. @dabrahams is correct in that our `block()` usage can cause issues if this is used as a general pattern.

Say a library has zlib as a dependency and uses `FetchContent` to declare it. If zlib is found via `find_package` one can use the `ZLIB::ZLIB` target without hesitation. However, if the dependent wants to inquire ZLIB's version using the `ZLIB_VERSION` variable, it will not be available due to the `block()` call enclosing `FetchContent_MakeAvailable`.

**Comment by wusatosi** [comment]
> I'm reopening this issue. @dabrahams is correct in that our `block()` usage can cause issues if this is used as a general pattern.
> 
> Say a library has zlib as a dependency and uses `FetchContent` to declare it. If zlib is found via `find_package` one can use the `ZLIB::ZLIB` target without hesitation. However, if the dependent wants to inquire ZLIB's version using the `ZLIB_VERSION` variable, it will not be available due to the `block()` call enclosing `FetchContent_MakeAvailable`.

Is there anything specific to exemplar we should be concerned/ fix?

**Comment by dabrahams** [comment]
I would follow [my example](https://github.com/hylo-lang/CMakeModules/blob/6577fca51495e82a94e76a8b70113c9e6b98f539/FetchHyloDependency.cmake#L12) if I were you, but fix the comment so it doesn't confuse people who know more CMake than I do.

**Comment by wusatosi** [comment]
@dabrahams @camio is there more potential discussion/ actions items in this issue thread? Otherwise we should close this.

**Comment by camio** [comment]
I agree. Removing the `block()` and `endblock()` calls will resolve this issue.
